### DISCUSSION TITLE
计划任务执行sh报错command not found

### DISCUSSION THREAD

**Discussion by JAINKRE** [1Panel-dev/1Panel#6914] [Q&A]
遇到一个奇怪的问题，机器上装了`node`和`npm`，用python写了一个自动化脚本，终端里运行正常，但放在计划任务`shell`运行却报环境丢失。

写了一个测试脚本：
```python
import os
os.system("pwd")
os.system("node -v")
os.system("python312 -V")
```
放在终端里运行没问题：
![image](https://github.com/user-attachments/assets/11fdab19-38e7-4da4-adae-0fa48d1991e9)

放在计划任务里错误：
![image](https://github.com/user-attachments/assets/867c95e8-b8f3-4ff4-9d8e-7be8aa027c34)

`python312`是本地安装的python3.12创建了软连接，为什么终端里python正常node就不正常？

现在定位不出来是哪里的问题，help!!!

**Comment by wanghe-fit2cloud** [comment]
先试一下全路径：
os.system("/path/to/python3.12 -V")

**Reply by JAINKRE** [reply]
解决了！感谢！看来是路径的问题。

```python
import os
os.system("pwd")
os.system("/usr/local/node/bin/node -v")
os.system("which python312")
```

终端:
![image](https://github.com/user-attachments/assets/cf8c8757-1341-46b0-b424-8a3b7c5fc34a)

计划任务：
![image](https://github.com/user-attachments/assets/e1cd6703-b696-416c-97e8-907201680ed2)

**Reply by JAINKRE** [reply]
虽然现在解决了，但问题是每台机器安装的路径并不一定完全相同。脚本里写绝对路径不是很好，只能先在脚本里用`which`动态获取，建议以后考虑可以优化一下这块。
